src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,467 @@
+  ctorSyntax : Array Syntax
+  /-- The flag that is `true` if the predicate was defined via `coinductive` keyword and `false`
+  otherwise. When we elaborate a mutual definition, we allow mixing `coinductive` and `inductive`
+  keywords, and hence we need to record this information.
+  -/
+  isGreatest : Bool
+  deriving Inhabited
+
+public def addFunctorPostfix : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str p (s ++ "_functor")
+  | Name.num p n => Name.num (addFunctorPostfix p) n
+
+public def removeFunctorPostfix : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str p (s.stripSuffix "_functor")
+  | Name.num p n => Name.num (removeFunctorPostfix p) n
+
+public def removeFunctorPostfixInCtor : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str (removeFunctorPostfix p) s
+  | Name.num p n => Name.num (removeFunctorPostfixInCtor p) n
+
+private def rewriteGoalUsingEq (goal : MVarId) (eq : Expr) (symm : Bool := false) : MetaM MVarId := do
+  let rewriteResult ← goal.rewrite (←goal.getType) eq symm
+  goal.replaceTargetEq rewriteResult.eNew rewriteResult.eqProof
+
+/--
+  Generates unrolling lemmas that relate coinductive predicates to their flat inductive forms.
+  These lemmas are essential for the constructor generation process, providing the bridge
+  between the user-facing coinductive predicates and their internal flat representations.
+-/
+private def generateEqLemmas (infos : Array InductiveVal) : MetaM Unit := do
+  let levels := infos[0]!.levelParams.map mkLevelParam
+  for info in infos do
+    let res ← forallTelescopeReducing info.type fun args _ => do
+      let params := args.take <| info.numParams - infos.size
+      let args := args.extract info.numParams
+
+      let lhs := mkConst (removeFunctorPostfix info.name) levels
+      let lhs := mkAppN lhs params
+      let lhs := mkAppN lhs args
+
+      let calls := infos.map fun info => mkAppN (mkConst (removeFunctorPostfix info.name) levels) params
+      let rhs := mkConst info.name levels
+      let rhs := mkAppN rhs (params ++ calls ++ args)
+
+      let goalType ← mkEq lhs rhs
+      let goal ← mkFreshExprMVar goalType
+
+      let goalMVarId := goal.mvarId!
+
+      let .some #[fixEq] ←  getEqnsFor? (removeFunctorPostfix info.name) | throwError "did not generate unfolding theorem"
+      let existentialEquiv := mkConst (info.name ++ `existential_equiv) levels
+
+      let mut fixEq := mkConst fixEq levels
+      fixEq := mkAppN fixEq params
+      for arg in args do
+        fixEq ← mkCongrFun fixEq arg
+
+      let newGoal ← rewriteGoalUsingEq goalMVarId existentialEquiv
+      newGoal.assign fixEq
+
+      let goal ← instantiateMVars goal
+      mkLambdaFVars (params ++ args) goal
+    trace[Elab.coinductive] "res: {res}"
+    addDecl <| .defnDecl (←mkDefinitionValInferringUnsafe
+      ((removeFunctorPostfix info.name) ++ `functor_unfold) info.levelParams (←inferType res) res .opaque)
```

Comment:
Minor style comment, but I find named parameters for `mkDefinitionValInferringUnsafe` helpful. Makes it look like the record it constructs, and I wouldn't be worried about confusing type and value.